multipartfile.go

```diff
@@ -125,102 +141,71 @@ func (it *multipartIterator) Node() Node {
 }
 
 func (it *multipartIterator) Next() bool {
-	if it.f.reader == nil {
+	if it.f.walker.reader == nil || it.err != nil {
 		return false
 	}
 	var part *multipart.Part
 	for {
-		var err error
-		part, err = it.f.reader.NextPart()
-		if err != nil {
-			if err == io.EOF {
-				return false
-			}
-			it.err = err
+		part, it.err = it.f.walker.getPart()
+		if it.err != nil {
+			return false
+		}
+
+		name := fileName(part)
+
+		// Is the file in a different directory?
+		if !strings.HasPrefix(name, it.f.path) {
 			return false
 		}
 
-		name, cf, err := newFileFromPart(it.f.fileName(), part, it.f.reader)
-		if err == ErrPartOutsideParent {
-			break
+		// Have we already entered this directory?
+		if it.curName != "" && strings.HasPrefix(name, path.Join(it.f.path, it.curName)) {
+			it.f.walker.consumePart()
+			continue
 		}
-		if err != ErrPartInChildTree {
-			it.curFile = cf
-			it.curName = name
-			it.err = err
-			return err == nil
+
+		// Make the path relative to the current directory.
+		name = strings.TrimLeft(name[len(it.f.path):], "/")
+
+		// Check if we need to create a fake directory (more than one
+		// path component).
+		if idx := strings.IndexByte(name, '/'); idx >= 0 {
+			it.curName = name[:idx]
+			it.curFile = &multipartDirectory{
+				path:   path.Join(it.f.path, it.curName),
+				walker: it.f.walker,
+			}
+			return true
 		}
-	}
+		it.curName = name
+
+		// Finally, advance to the next file.
+		it.curFile, it.err = nextFile(it.f.walker)
 
-	it.err = it.f.reader.put(part)
-	return false
+		return it.err == nil
+	}
 }
 
 func (it *multipartIterator) Err() error {
+	if it.err == io.EOF {
```

Comment:
I'm using this as a "done" signal. Probably not necessary but I'm paranoid.

Reply:
A short comment would be nice